### PR TITLE
FIX-JENKINS-40605_Time_reported_is_not_correct

### PR DESCRIPTION
**Pull request by scherler**
# Description

See [JENKINS-40605](https://issues.jenkins-ci.org/browse/JENKINS-40605).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@reviewbybees

### CODE REVIEW

File: blueocean-web/src/main/js/config.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ export default class Config {
         this._rootURL = options.rootURL || '';
         this._resourceURL = options.resourceURL || '';
         this._adjunctURL = options.adjunctURL || '';
+        this._timeOffset = options.timeOffset || '';
```
Comment (tfennelly) [inline blueocean-web/src/main/js/config.js]: How about using a more descriptive name e.g. `serverBrowserTimeSkewMillis` ?

**Comment by tfennelly** [comment]
Looking good so far @scherler.

File: blueocean-web/src/main/js/config.js (hunk 1 above)
Comment (scherler) [inline blueocean-web/src/main/js/config.js]: Makes sense. Will change it.

**Comment by michaelneale** [comment]
nice work

File: blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx [hunk 2]
```diff
@@ -2,11 +2,27 @@
 
 import React, { Component, PropTypes } from 'react';
 import { Icon } from '@jenkins-cd/react-material-icons';
+import { logging } from '@jenkins-cd/blueocean-core-js';
 import { ExpandablePath, ReadableDate, LiveStatusIndicator, TimeDuration } from '@jenkins-cd/design-language';
 import ChangeSetToAuthors from './ChangeSetToAuthors';
-import moment from 'moment';
+import { TimeManager } from '../util/serverBrowserTimeHarmonize';
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: mmmm synergy

File: blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx [hunk 3]
```diff
@@ -2,11 +2,27 @@
...
+
+const logger = logging.logger('io.jenkins.blueocean.dashboard');
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: I'd be inclined to include `RunDetailsHeader` in the category name.

File: blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx [hunk 4]
```diff
@@ -34,13 +50,25 @@ class RunDetailsHeader extends Component {
...
+        const isRunning = () => run.isRunning() || run.isPaused();
+        const {
+            durationMillis,
+            endTime,
+            startTime,
+        } = timeManager.harmonizeTimes({
+            endTime: run.endTime,
+            startTime: run.startTime,
+            durationInMillis: run.durationInMillis,
+        }, skewMillis);
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: Could you just pass the `run` instance to the `harmonizeTimes` function?

```javascript
const {
    durationMillis,
    endTime,
    startTime,
} = timeManager.harmonizeTimes(run, skewMillis);
```

Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: And (NIT :) ) I don't get the benefit of the destruction here. How is all that more clear than a simple ...

```javascript
const harmonizedTimes = timeManager.harmonizeTimes(run, skewMillis);
logger.debug('rundetails times:', harmonizedTimes);
```

File: blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * @param run
+ * @param skewMillis
+ * @returns {{ durationMillis: diff, endTime: Date, startTime: Date}}
+ */
+export class TimeManager {
+    currentTime() {
+        return moment();
+    }
+
+    harmonizeTimes(run, skewMillis) {
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: Is the param named correctly here? I'd have thought it should be named something other than `run`. Any oject containing `startTime`, `endTime` an `durationMillis`, no?

Comment (tfennelly) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: AAMOI... if you have `startTime` and `endTime`, why is `durationMillis` needed?

File: blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            serverStartTime.add({ milliseconds: Math.abs(skewMillis) });
+        } else {
+            serverStartTime.subtract({ milliseconds: skewMillis });
+        }
+    // export the harmonized start time
+        const startTime = serverStartTime.toJSON();
+    // how long has passed from the start to now in milliseconds
+        const timeElapsed = clientTime.diff(serverStartTime, 'milliseconds');
+    // assume we do not have an end date
+        let endTime = null;
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: Is there a reason why the comments are all out of line with the code?

File: blueocean-web/src/main/js/config.js [hunk 7]
```diff
@@ -10,6 +10,7 @@ export default class Config {
         this._rootURL = options.rootURL || '';
         this._resourceURL = options.resourceURL || '';
         this._adjunctURL = options.adjunctURL || '';
+        this._serverBrowserTimeSkewMillis = options.timeOffset || '';
```
Comment (tfennelly) [inline blueocean-web/src/main/js/config.js]: `timeOffset` the right name for that?

File: blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx (hunk 4 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: well, yes and now. We not always use the run but as well other object, which do not follow that pattern. However, I could harmonize that though.

File: blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx (hunk 3 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/RunDetailsHeader.jsx]: `io.jenkins.blueocean.dashboard.RunDetailsHeader` you mean? Does that mean I have to set the debug level for that specific component or do I get  everything "under" `'io.jenkins.blueocean.dashboard'`

File: blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js (hunk 5 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: correct

File: blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js (hunk 6 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: Will fix that, I guess some IDE went nuts

File: blueocean-web/src/main/js/config.js (hunk 7 above)
Comment (scherler) [inline blueocean-web/src/main/js/config.js]: well we could use `serverBrowserTimeSkewMillis` all over the place

File: blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js (hunk 6 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/util/serverBrowserTimeHarmonize.js]: Actually it is a combination of IDE formatting and then afterwards using lint:fix. Seems comments are not being intended like normal code.

**Pull request merged**